README
				Fake64
				======

What is fake64?
---------------

Fake64 is an n64 emulator for linux which is currently in the early stages of
development. However, most of the basic features are fully implemented and many
more advanced ones such as 3d graphics are at least partially in place.

It is licensed under the GPL.

What do i need to run it?
-------------------------

Fake64 requires a reasonably modern linux system with the following:

	X, XFree86 => 4.0.0 with DRI recommended (http://www.xfree86.org)
	A GL implementation, usually provided with X
	A GLU implementation, usually provided with X
	SDL, => 1.1.0 recommended (http://www.libsdl.org)
	All the usual compiling tools, cc, ld, etc ...

Additionally, if you are building from cvs you will require automake and
autoconf.

From a hardware point of view, fake64 will run on a wide variety of hardware
specs. At the moment, the limitations on hardware are an x86
system with mmx support. Obviously the faster your computer, the faster fake64
will go, but you don't need a supercomputer to see reasonable results.

The most important factor affecting performance, is hardware accelleration.
Getting hardware accelleration set up in linux is a complex subject due to the
wide variety of hardware available. A good resource for the most common cards
is available at the dri project website (http://dri.sourceforge.net).

Where can I get it?
-------------------

You can obtain fake64 in various ways from our website, which currently resides
at http://insomniac.ath.cx:81/~bluefyre/

Okay, I've downloaded it, now what?
-----------------------------------

For instructions on installing fake64, please read the file INSTALL in this
directory.

Some non-commercial demo roms are distributed with fake64 in the file
testroms.tar.bz2, which can be extracted with the command:

	tar xvIf testroms.tar.bz2

Please note that the authors of fake64 in no way encourage you to break the law
by illegally obtaining commercial roms, we give you a stick, but we don't make
you hit people with it, or something.

For more information on what roms fake64 can run, see the compatability list
on the site.

How can I customise fake64?
---------------------------

Fake64 contains a flexible runtime configuration system. It reads commands from
a file ~/.fake64rc

For examples of what you can put in this file, see sample.fake64rc.

Please note that at present, the configuration parsing isn't particularly
resilient, so if you try and do funny things with formatting, expect segfaults.

Here is a list of some of the most common things to do in a fake64rc:

	SetDefault - Establishes a default section to run for all roms
	SetRom - Establishes a section to run for an individual rom file

	Section - Defines a section
	Include - Includes a section
	LoadModule - Loads a module
	Debug - Sets a debug level, only if compiled --enable-debug
	Option - sends an option

Option is probably the most used command, and options are generally formatted:

	Option [module] [option] [value]

For example:

	Option video 2d 0

To turn off 2d.

Most options can be given either a 0 or 1 as a value to turn the feature on or
off. Some accept any number, or take several values.

Here are some commonly used options:

	video:
		width [number] - set window width
		height [number] - set window height
		fpslimiter [0|1] - turn on fps limiter (a little iffy)
		fpslimit [number] - set fps limit
		2d [0|1] - set 2d on or off

	cpu:
		run [0|1] - start immediately?
		cf [number] - set counter factor (1-10 recommended)
		patch [address] [value] - apply patch
		patches [0|1] - activate patches

	input:
		mapkey [controller] [button] [scancode] - map a key
		resetmaps - reset all keymaps
		mapstick[x|y] [controller] [value] [scancode] - map to analogue
		controllers [1-8] - number of controllers active



ChangeLog
Oct 29 2001 by HCl:

    made gfx update every 625000 instructions -> 1964 demo fully works
    (isn't the right way to do it, but it works)

Oct 20 2001 by Bluefyre:

    Made a sexy package type thinger

Oct 19 2001 by HCl:

    Changed romloader quite a bit: modules are loaded interactively.

Oct 03 2001 by Bluefyre:

    Realised the named stuff is worse than useless unless the debugger dumps
    stuff with those names too, so i made it do that.

    Fixed a bug in my symlink makefile thing

    Hacked with the romloader so you can run it from cpucore_pure_c/ now
    without it complaining it can't find the libs

Oct 02 2001 by Bluefyre:

    Miscellaneousness, I made it make the symlink to loader automatically

    The debugger now shows you the register names not numbers

Aug 22 2001 by HCl:

    (Note that we haven't been able to do a lot because the cvs server has 
     been down)

    Additions:

    cpucore_pure_c:
        
    a crude debugger :))) (libcpu.c, maybe move to it's own sourcefile)

    commands: run next breakpoint show dumpregs help quit

Jul 23/24 2001 by HCl:

    Additions:

    Lots of store/load instructions, and a few others
    massive speed optimizations in store/load instructions -> code much faster
    optimized CFLAGS

Jul 21/22 2001 by HCl:

    Additions/Optimizations:
 
    cpucore_pure_c:

    added more ram areas
    added lots of other instructions
    added dCPU_COP1 support
    added dCPU_COP1_CFC1
    added dCPU_COP1_CTC1 
    added eCPU_COP1_CFC1
    added eCPU_COP1_CTC1
    added eCPU_LW
    added fpr registers of cop1 to registers.h
    optimized eCPU_SW
    added CPUdelay slot (omg i can't believe we forgot this)
	-> code makes much more sense

    Changes:

    cpucore_pure_c:

    fixed comments in registers.h
    fixed ADDIU in exec.c: sgi people use misleading names for opcodes!

Jul 21 2001 by Bluefyre:

    Additions:

    cpucore_pure_c:

    added eCPU_COP0 support
    added MFC0/MTC0 (cop0) (cop0_scp.c)
    added COP0 (exec.c)
    changed some .h's to match opcodes.h
    changed opstruct.c to match

    Changes:

    cpucore_pure_c:

    made do_a_dump() dump fgrs (libcpu.c)
    made do_a_dump() print to stderr (libcpu.c)
    messed with do_a_dump() to make it look nicer (libcpu.c)

    made a debug_do_a_dump() for HCl (libcpu.c)
    made EXCESSIVE_DEBUG use debug_do_a_dump() (libcpu.c)

    changed the Changelog :O)

Jul 18/19 2001 by HCl: 

    Additions/Optimizations:

    cpucore_pure_c:
 
    added MFC0/MTC0 (cop0) (disass.c)
    added JAL/JR (exec.c)
    optimized SLTU (exec.c)
    optimized BEQ (exec.c)
    optimized BNE (exec.c
    added ADDI (exec.c)
    added RDRAM memory support (experimental) (libcpu.c,exec.c)
    put ifdef checks around info functions in main_cpu_loop (libcpu.c)
    added mentioning the PC when disassembling an instruction (libcpu.c)
    Error handling optimized (libcpu.c)
		  
    Fixes:
  
    romloader:
  
    load_module.h missing  -> fixed
    bad,sloppy asm code in fix_rom_name -> fixed: replaced with C code 	

    cpucore_pure_c:

    counting of instructions disassembled -> fixed
    instruction pointer didn't use native addys -> jumps broken -> fixed
    instruction pointer advanced after jump -> jumps broken -> fixed
    eCPU_BEQ, eCPU_BNE didn't add 4 to the jump addy -> fixed
    dCPU_JAL* : didn't shift target left by 2 -> fixed


include/memory.h
#ifndef MEMORY_H
#define MEMORY_H

#define RAM_SIZE	0x00800000ul
#if (RAM_SIZE > 0x20000000ul)
#error Cannot allocate this much RDRAM with the RCP memory model.
#endif

/*
 * Unless the pre-processor receives instructions to inaccurately emulate the
 * byte order for speed purposes, we'll store Nintendo 64 DRAM accurately.
 * This can actually be faster in some cases than aliasing 32-bit writes in
 * client-native endian, anyway.
 */
#ifndef CLIENT_ENDIAN
#define SERVER_ENDIAN
#endif

#ifdef CLIENT_ENDIAN
#define BYTE_ADDRESS_SWAP       3
/* should stay 7 for 64-bit little-endian, 3 for 32-bit, etc. */
#else
#define BYTE_ADDRESS_SWAP       0
#endif
#define HALFWORD_ADDRESS_SWAP   ((BYTE_ADDRESS_SWAP) / 2)
#define WORD_ADDRESS_SWAP       ((BYTE_ADDRESS_SWAP) / 4)
#define DOUBLEWORD_ADDRESS_SWAP ((BYTE_ADDRESS_SWAP) / 8)

#define BES(x)  ((x) ^ ((BYTE_ADDRESS_SWAP) & ~0))
#define HES(x)  ((x) ^ ((BYTE_ADDRESS_SWAP) & ~1))
#define WES(x)  ((x) ^ ((BYTE_ADDRESS_SWAP) & ~3))
#define DES(x)  ((x) ^ ((BYTE_ADDRESS_SWAP) & ~7))


#define DRAM_BASE_REG	0x03F00000
#define SP_DMEM		0x04000000
#define SP_DMEM_END	0x04000FFF
#define SP_IMEM		0x04001000
#define SP_IMEM_END	0x04001FFF
#define SP_BASE_REG	0x04040000
#define DPC_BASE_REG	0x04100000
#define DPS_BASE_REG	0x04200000
#define MI_BASE_REG	0x04300000
#define VI_BASE_REG	0x04400000
#define AI_BASE_REG	0x04500000
#define PI_BASE_REG	0x04600000
#define RI_BASE_REG	0x04700000
#define SI_BASE_REG	0x04800000
#define ROM_BASE	0x10000000
#define PIF_MEM_BASE	0x1FC00000

extern uint8 *RDREGS;		// RDRAM registers
extern uint8 *SPDIMEM;
extern uint8 *SPREGS;		// SP registers
extern uint8 *SP_REG;
extern uint8 *DPREGS;		// DP registers
extern uint8 *DPSREGS;		// DP span registers
extern uint8 *MIREGS;		// MIPS Interface registers
extern uint8 *VIREGS;		// Video Interface registers
extern uint8 *AIREGS;		// Audio Interface registers
extern uint8 *PIREGS;		// Peripheral Interface registers
extern uint8 *RIREGS;		// RDRAM Interface registers
extern uint8 *SIREGS;		// Serial Interface registers
extern uint8 *PIFMEM;		// PIF area
extern uint8 *RDRAM;		// rdram
extern uint8 *ROM;		// rom(header&everythin)

extern void* RAM_OFFSET_MAP[];

void* Check_Load(int32 addr);
void* Check_Store(int32 addr, uint32 value);

#include "romheader.h"
int alloc_memory (struct rom *rom);
void init_memory (int romlength);

#define SI_BASE_REG             0x04800000
#define SI_DRAM_ADDR_REG        (SI_BASE_REG+0x00)
#define SI_PIF_ADDR_RD64B_REG   (SI_BASE_REG+0x04)
#define SI_PIF_ADDR_WR64B_REG   (SI_BASE_REG+0x10)
#define SI_STATUS_REG           (SI_BASE_REG+0x18)

#define SI_STATUS_DMA_BUSY      0x0001
#define SI_STATUS_RD_BUSY       0x0002
#define SI_STATUS_DMA_ERROR     0x0008
#define SI_STATUS_INTERRUPT     0x1000

#define PI_BASE_REG             0x04600000
#define PI_DRAM_ADDR_REG        (PI_BASE_REG+0x00)
#define PI_CART_ADDR_REG        (PI_BASE_REG+0x04)
#define PI_RD_LEN_REG           (PI_BASE_REG+0x08)
#define PI_WR_LEN_REG           (PI_BASE_REG+0x0C)
#define PI_STATUS_REG           (PI_BASE_REG+0x10)
#define PI_BSD_DOM1_LAT_REG     (PI_BASE_REG+0x14)
#define PI_BSD_DOM1_PWD_REG     (PI_BASE_REG+0x18)
#define PI_BSD_DOM1_PGS_REG     (PI_BASE_REG+0x1C)
#define PI_BSD_DOM1_RLS_REG     (PI_BASE_REG+0x20)
#define PI_BSD_DOM2_LAT_REG     (PI_BASE_REG+0x24)
#define PI_BSD_DOM2_PWD_REG     (PI_BASE_REG+0x28)
#define PI_BSD_DOM2_PGS_REG     (PI_BASE_REG+0x2C)
#define PI_BSD_DOM2_RLS_REG     (PI_BASE_REG+0x30)

#define PI_DOMAIN1_REG          PI_BSD_DOM1_LAT_REG
#define PI_DOMAIN2_REG          PI_BSD_DOM2_LAT_REG

#define PI_DOM_LAT_OFS          0x00
#define PI_DOM_PWD_OFS          0x04
#define PI_DOM_PGS_OFS          0x08
#define PI_DOM_RLS_OFS          0x0C

#define PI_STATUS_ERROR         0x04
#define PI_STATUS_IO_BUSY       0x02
#define PI_STATUS_DMA_BUSY      0x01

#define PI_STATUS_RESET         0x01
#define PI_SET_RESET            PI_STATUS_RESET

#define PI_STATUS_CLR_INTR      0x02
#define PI_CLR_INTR             PI_STATUS_CLR_INTR

#define PI_DMA_BUFFER_SIZE      128

#define PI_DOM1_ADDR1           0x06000000
#define PI_DOM1_ADDR2           0x10000000
#define PI_DOM1_ADDR3           0x1FD00000
#define PI_DOM2_ADDR1           0x05000000
#define PI_DOM2_ADDR2           0x08000000

#define MI_BASE_REG             0x04300000
#define MI_INIT_MODE_REG        (MI_BASE_REG+0x00)
#define MI_MODE_REG             MI_INIT_MODE_REG

#define MI_CLR_INIT             0x0080
#define MI_SET_INIT             0x0100
#define MI_CLR_EBUS             0x0200
#define MI_SET_EBUS             0x0400
#define MI_CLR_DP_INTR          0x0800
#define MI_CLR_RDRAM            0x1000
#define MI_SET_RDRAM            0x2000

#define MI_MODE_INIT            0x0080
#define MI_MODE_EBUS            0x0100
#define MI_MODE_RDRAM           0x0200

#define MI_VERSION_REG          (MI_BASE_REG+0x04)
#define MI_NOOP_REG             MI_VERSION_REG
#define MI_INTR_REG             (MI_BASE_REG+0x08)
#define MI_INTR_REG_R		*((uint32 *)(MIREGS+0x08))
#define MI_INTR_MASK_REG        (MI_BASE_REG+0x0C)
#define MI_INTR_MASK_REG_R	*((uint32 *)(MIREGS+0x0C))

#define MI_INTR_SP              0x01
#define MI_INTR_SI              0x02
#define MI_INTR_AI              0x04
#define MI_INTR_VI              0x08
#define MI_INTR_PI              0x10
#define MI_INTR_DP              0x20

#define MI_INTR_MASK_CLR_SP     0x0001
#define MI_INTR_MASK_SET_SP     0x0002
#define MI_INTR_MASK_CLR_SI     0x0004
#define MI_INTR_MASK_SET_SI     0x0008
#define MI_INTR_MASK_CLR_AI     0x0010
#define MI_INTR_MASK_SET_AI     0x0020
#define MI_INTR_MASK_CLR_VI     0x0040
#define MI_INTR_MASK_SET_VI     0x0080
#define MI_INTR_MASK_CLR_PI     0x0100
#define MI_INTR_MASK_SET_PI     0x0200
#define MI_INTR_MASK_CLR_DP     0x0400
#define MI_INTR_MASK_SET_DP     0x0800

#define MI_INTR_MASK_SP         0x01
#define MI_INTR_MASK_SI         0x02
#define MI_INTR_MASK_AI         0x04
#define MI_INTR_MASK_VI         0x08
#define MI_INTR_MASK_PI         0x10
#define MI_INTR_MASK_DP         0x20

#define RI_BASE_REG             0x04700000
#define RI_MODE_REG             (RI_BASE_REG+0x00)
#define RI_CONFIG_REG           (RI_BASE_REG+0x04)
#define RI_CURRENT_LOAD_REG     (RI_BASE_REG+0x08)
#define RI_SELECT_REG           (RI_BASE_REG+0x0C)
#define RI_REFRESH_REG          (RI_BASE_REG+0x10)
#define RI_COUNT_REG            RI_REFRESH_REG
#define RI_LATENCY_REG          (RI_BASE_REG+0x14)
#define RI_RERROR_REG           (RI_BASE_REG+0x18)
#define RI_WERROR_REG           (RI_BASE_REG+0x1C)

#define SP_BASE_REG             0x04040000
#define SP_MEM_ADDR_REG         (SP_BASE_REG+0x00)
#define SP_DRAM_ADDR_REG        (SP_BASE_REG+0x04)
#define SP_RD_LEN_REG           (SP_BASE_REG+0x08)
#define SP_WR_LEN_REG           (SP_BASE_REG+0x0C)
#define SP_STATUS_REG           (SP_BASE_REG+0x10)
#define SP_DMA_FULL_REG         (SP_BASE_REG+0x14)
#define SP_DMA_BUSY_REG         (SP_BASE_REG+0x18)
#define SP_SEMAPHORE_REG        (SP_BASE_REG+0x1C)
#define SP_SEMAPHORE_REG        (SP_BASE_REG+0x1C)
#define SP_PC_REG               0x04080000
#define SP_DMA_DMEM             0x0000
#define SP_DMA_IMEM             0x1000 

#define SP_CLR_HALT             0x00001     /* Bit  0: clear halt */
#define SP_SET_HALT             0x00002     /* Bit  1: set halt */
#define SP_CLR_BROKE            0x00004     /* Bit  2: clear broke */
#define SP_CLR_INTR             0x00008     /* Bit  3: clear intr */
#define SP_SET_INTR             0x00010     /* Bit  4: set intr */
#define SP_CLR_SSTEP            0x00020     /* Bit  5: clear sstep */
#define SP_SET_SSTEP            0x00040     /* Bit  6: set sstep */
#define SP_CLR_INTR_BREAK       0x00080     /* Bit  7: clear intr on break */
#define SP_SET_INTR_BREAK       0x00100     /* Bit  8: set intr on break */
#define SP_CLR_SIG0             0x00200     /* Bit  9: clear signal 0 */
#define SP_SET_SIG0             0x00400     /* Bit 10: set signal 0 */
#define SP_CLR_SIG1             0x00800     /* Bit 11: clear signal 1 */
#define SP_SET_SIG1             0x01000     /* Bit 12: set signal 1 */
#define SP_CLR_SIG2             0x02000     /* Bit 13: clear signal 2 */
#define SP_SET_SIG2             0x04000     /* Bit 14: set signal 2 */
#define SP_CLR_SIG3             0x08000     /* Bit 15: clear signal 3 */
#define SP_SET_SIG3             0x10000     /* Bit 16: set signal 3 */
#define SP_CLR_SIG4             0x20000     /* Bit 17: clear signal 4 */
#define SP_SET_SIG4             0x40000     /* Bit 18: set signal 4 */
#define SP_CLR_SIG5             0x80000     /* Bit 19: clear signal 5 */
#define SP_SET_SIG5            0x100000     /* Bit 20: set signal 5 */
#define SP_CLR_SIG6            0x200000     /* Bit 21: clear signal 6 */
#define SP_SET_SIG6            0x400000     /* Bit 22: set signal 6 */
#define SP_CLR_SIG7            0x800000     /* Bit 23: clear signal 7 */
#define SP_SET_SIG7           0x1000000     /* Bit 24: set signal 7 */

#define SP_STATUS_HALT          0x001
#define SP_STATUS_BROKE         0x002
#define SP_STATUS_DMA_BUSY      0x004
#define SP_STATUS_DMA_FULL      0x008
#define SP_STATUS_IO_FULL       0x010         
#define SP_STATUS_SSTEP         0x020        
#define SP_STATUS_INTR_BREAK    0x040          
#define SP_STATUS_SIG0          0x080           
#define SP_STATUS_SIG1          0x100         
#define SP_STATUS_SIG2          0x200         
#define SP_STATUS_SIG3          0x400         
#define SP_STATUS_SIG4          0x800      
#define SP_STATUS_SIG5         0x1000           
#define SP_STATUS_SIG6         0x2000          
#define SP_STATUS_SIG7         0x4000        

#define AI_DRAM_ADDR_REG        (AI_BASE_REG+0x00)
#define AI_LEN_REG              (AI_BASE_REG+0x04)
#define AI_CONTROL_REG          (AI_BASE_REG+0x08)
#define AI_STATUS_REG           (AI_BASE_REG+0x0C)
#define AI_DACRATE_REG          (AI_BASE_REG+0x10)
#define AI_BITRATE_REG          (AI_BASE_REG+0x14)
#define AI_CONTROL_DMA_ON       0x01
#define AI_CONTROL_DMA_OFF      0x00
#define AI_STATUS_FIFO_FULL     0x80000000
#define AI_STATUS_DMA_BUSY      0x40000000
#define AI_MAX_DAC_RATE         16384
#define AI_MIN_DAC_RATE         132
#define AI_MAX_BIT_RATE         16 
#define AI_MIN_BIT_RATE         2
#define AI_NTSC_MAX_FREQ        368000
#define AI_NTSC_MIN_FREQ        3000
#define AI_PAL_MAX_FREQ         376000
#define AI_PAL_MIN_FREQ         3050
#define AI_MPAL_MAX_FREQ        368000
#define AI_MPAL_MIN_FREQ        3000

#define VI_STATUS_REG		(VI_BASE_REG)
#define VI_CONTROL_REG		VI_STATUS_REG
#define VI_ORIGIN_REG		(VI_BASE_REG+0x04)
#define VI_DRAM_ADDR_REG	VI_ORIGIN_REG
#define VI_WIDTH_REG		(VI_BASE_REG+0x08)
#define VI_H_WIDTH_REG		VI_WIDTH_REG
#define VI_INTR_REG		(VI_BASE_REG+0x0C)
#define VI_V_INTR_REG		VI_INTR_REG
#define VI_CURRENT_REG		(VI_BASE_REG+0x10)
#define VI_V_CURRENT_LINE_REG	VI_CURRENT_REG
#define VI_BURST_REG		(VI_BASE_REG+0x14)
#define VI_TIMING_REG		VI_BURST_REG
#define VI_V_SYNC_REG		(VI_BASE_REG+0x18)
#define VI_H_SYNC_REG		(VI_BASE_REG+0x1C)
#define VI_LEAP_REG		(VI_BASE_REG+0x20)
#define VI_H_SYNC_LEAP_REG	VI_LEAP_REG
#define VI_H_START_REG		(VI_BASE_REG+0x24)
#define VI_H_VIDEO_REG		VI_H_START_REG
#define VI_V_START_REG		(VI_BASE_REG+0x28)
#define VI_V_VIDEO_REG		VI_V_START_REG
#define VI_V_BURST_REG		(VI_BASE_REG+0x2C)
#define VI_X_SCALE_REG		(VI_BASE_REG+0x30)
#define VI_Y_SCALE_REG		(VI_BASE_REG+0x34)
#endif
